src/ragger/navigator/interface.py

```diff
@@ -0,0 +1,288 @@
+                return True
+            now = time()
+        return False
+
+    def _compare_snaps(self, path: Path, test_case_name: Path, last_img_idx: int):
+        snaps_golden_path = self._get_snaps_dir_path(path, test_case_name, True)
+        snaps_tmp_path = self._get_snaps_dir_path(path, test_case_name, False)
+
+        if not snaps_golden_path.is_dir():
+            raise ValueError(f"Golden snapshots directory ({snaps_golden_path}) does not exist.")
+
+        if not snaps_tmp_path.is_dir():
+            raise ValueError(f"Temporary snapshots directory ({snaps_golden_path}) does not exist.")
+
+        for i in range(0, last_img_idx + 1):
+            golden = self._get_snap_path(snaps_golden_path, i)
+            tmp = self._get_snap_path(snaps_tmp_path, i)
+            if not screenshot_equal(str(golden), str(tmp)):
+                raise ValueError(f"Screenshots {tmp} does not match golden.")
+
+    @abstractmethod
+    def navigate(self, instructions: List[NavigationInstruction]) -> None:
+        """
+        Navigate on the device according to a set of navigation instructions provided.
+
+        :param instructions: Set of navigation instructions.
+        :type instructions: List[NavigationInstruction]
+
+        :raises NotImplementedError: If the navigation instruction is not implemented.
+
+        :return: None
+        :rtype: NoneType
+        """
+        raise NotImplementedError
+
+    def navigate_and_compare(self,
+                             path: Path,
+                             test_case_name: Path,
+                             instructions: List[NavigationInstruction],
+                             first_instruction_wait: float = 0.5,
+                             middle_instruction_wait: float = 0.1,
+                             last_instruction_wait: float = 0.5) -> None:
+        """
+        Navigate on the device according to a set of navigation instructions
+        provided then compare each step snapshot with "golden images".
+
+        :param path: Absolute path to the snapshots directory.
+        :type path: Path
+        :param test_case_name: Relative path to the test case snapshots directory (from path).
+        :type test_case_name: Path
+        :param instructions: List of navigation instructions.
+        :type instructions: List[NavigationInstruction]
+        :param first_instruction_wait: Sleeping time before the first snapshot
+        :type first_instruction_wait: float
+        :param middle_instruction_wait: Sleeping time between between each navigation instruction
+        :type middle_instruction_wait: float
+        :param last_instruction_wait: Sleeping time after the last navigation instruction
+        :type last_instruction_wait: float
+
+        :raises ValueError: If one of the snapshots do not match.
+
+        :return: None
+        :rtype: NoneType
+        """
+        if not isinstance(self._backend, SpeculosBackend):
```

Comment:
Not found of this. The whole point of the backends is to avoid these discriminated treatments. if we start splitting the usage depending on the backend type, the code will soon become illegible condition hell.
I'd rather accept that code run uselessly - but won't crash, all methods should be defined and correct - on some backend. **For now**. At some point, we will need those empty functions and replace them with more meaningful code - like ask a user to check if some golden image fits what he's seeing on the physical device screen.

Reply:
+1
And Navigation DOES make sense in LedgerComm and LedgerWallet, even if it's just asking the tester `Press right button once` and `Ensure that this image corresponds to the current Nano screen`

Reply:
OR !

As discussed with @agrojean-ledger we could:
- add a `parser.addoption("--golden_snapshot")`
- pass it in `SpeculosBackend` instantiation `kwargs`
- have a specific behavior in theses cases which instead of comparing the snapshot it saves it in the golden dir.

with such behavior, we could go back to comparing the snapshot after each instruction, which would allow to do it on `LedgerComm` and `LedgerWallet`.

Reply:
I suggest we postpone this to another PR, we'll see along the way

Reply:
I've update the PR with an additional commit on which we can hopefully find a consensus!